### PR TITLE
Add keep_if and keep_unless

### PR DESCRIPTION
**Pull request by xvw**

### CODE REVIEW

**Review by Fenntasy — APPROVED** [review]

**Review by arnaudmorisset — APPROVED** [review]

File: lib/kaur/result.ex [hunk 1]
```diff
@@ -91,6 +91,52 @@ defmodule Kaur.Result do
...
+      iex> res = Kaur.Result.ok(10)
+      ...> Kaur.Result.keep_if(res, &(&1 > 10), "must be > of 10")
+      {:error, "must be > of 10"}
+
+      iex> res = Kaur.Result.error(:no_value)
+      ...> Kaur.Result.keep_if(res, &(&1 > 10), "must be > of 10")
+      {:error, :no_value}
+  """
+  @spec keep_if(result_tuple, (any -> boolean), any) :: result_tuple
+  def keep_if(result, predicate, error_message \\ :invalid)
```
Comment (kdisneur) [inline lib/kaur/result.ex:111]: I don't understand when it's going to be useful to have this default error message?

File: lib/kaur/result.ex [hunk 2]
```diff
@@ -91,6 +91,52 @@ defmodule Kaur.Result do
...
+      iex> res = Kaur.Result.ok([1])
+      ...> Kaur.Result.keep_unless(res, &Enum.empty?/1)
+      {:ok, [1]}
+
+      iex> res = Kaur.Result.ok([])
+      ...> Kaur.Result.keep_unless(res, &Enum.empty?/1, "list cannot be empty")
+      {:error, "list cannot be empty"}
+  """
+  @spec keep_unless(result_tuple, (any -> boolean), any) :: result_tuple
+  def keep_unless(result, predicate, error_message \\ :invalid) do
```
Comment (kdisneur) [inline lib/kaur/result.ex]: Which naming do you prefer? `keep_unless` or `reject_if`?

Comment (xvw) [reply lib/kaur/result.ex]: In fact, I prefer `reject_if`, but IMHO, keep_unless is "more" in the if/unless logic ... What's your point ?

File: lib/kaur/result.ex (hunk 1 above)
Comment (Fenntasy) [reply lib/kaur/result.ex:111]: When you use it without any interest for an error message.
That was the use case of @arnaudmorisset

File: lib/kaur/result.ex (hunk 2 above)
Comment (kdisneur) [reply lib/kaur/result.ex]: We have `filter` and `reject` in `Enum` when we want to keep or reject things so I was wondering if we should follow this same naming or not.

I don't have real opinion, just asking what would be the best naming

Comment (xvw) [reply lib/kaur/result.ex]: Ok, agreed ;)

**Review by kdisneur — APPROVED** [review]

**Pull request merged**